zict/lru.py

```diff
@@ -82,11 +94,18 @@ def evict(self):
         w: weight
         """
         k, priority = self.heap.popitem()
+        v = self.d.pop(k)
+        try:
+            for cb in self.on_evict:
+                cb(k, v)
+        except Exception:
+            # e.g. if a callback tried storing to disk and raised a disk full error
+            self.heap[k] = priority
+            self.d[k] = v
+            raise
```

Comment:
I think there is a problem with this raise. I was trying some things around where due I ended up having on fast multiple keys as a result of disk_limit being reached, that I'll write in priority order {'a': 400, 'd': 100, 'e': 500} . In slow after deleting a key I have only {'c': 200}, let's say the  disk_limit =  500 . If I add a new key ('f': 100) that is small to go on fast, I'd expect 'd' that is on fast to be moved to slow. But this doesn't happen since the first key that it tries to move due to low priority is 'e' and it's too big that it raises and therefore it stops going over the rest of the dictionary, even though there is clearly room for 'd' on slow.

Reply:
That's fine. The LRU does not and should not know that the reason for the exception is that the data is too large. It would be expensive and complicated to try writing the other keys just in case they do not fail.  Also, I don't think the use case is very interesting in real life in distributed.